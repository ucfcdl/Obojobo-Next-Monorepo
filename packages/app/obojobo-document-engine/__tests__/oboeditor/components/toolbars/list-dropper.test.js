import { mount, shallow } from 'enzyme'
import React from 'react'

import ListDropper from '../../../../src/scripts/oboeditor/components/toolbars/list-dropper'

const LIST_NODE = 'ObojoboDraft.Chunks.List'

jest.useFakeTimers()

describe('List Dropper', () => {
	test('List Dropper node', () => {
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = shallow(<ListDropper type="unordered" bullets={bullets} />)
		const tree = component.html()
		expect(tree).toMatchSnapshot()
	})

	test('List Dropper toggles bullet', () => {
		const editor = {
			changeToType: jest.fn(),
			children: [
				{
					type: 'mock-type',
					children: [{ text: '' }]
				}
			],
			selection: {
				anchor: { path: [0], offset: 0 },
				focus: { path: [0], offset: 0 }
			},
			isVoid: () => false,
			isInline: () => false
		}

		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = mount(<ListDropper type="ordered" bullets={bullets} editor={editor} />)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(editor.changeToType).toHaveBeenCalled()
	})

	test('List Dropper toggles bullet on list', () => {
		const editor = {
			changeToType: jest.fn(),
			children: [
				{
					type: LIST_NODE,
					content: { listStyles: { type: 'unordered' } },
					children: [{ text: '' }]
				}
			],
			selection: {
				anchor: { path: [0], offset: 0 },
				focus: { path: [0], offset: 0 }
			},
			isVoid: () => false,
			isInline: () => false
		}

		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = mount(<ListDropper type="ordered" bullets={bullets} editor={editor} />)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(editor.changeToType).toHaveBeenCalled()
	})

	test('List Dropper toggles bullet to text', () => {
		const editor = {
			changeToType: jest.fn(),
			children: [
				{
					type: LIST_NODE,
					content: { listStyles: { type: 'ordered' } },
					children: [{ text: '' }]
				}
			],
			selection: {
				anchor: { path: [0], offset: 0 },
				focus: { path: [0], offset: 0 }
			},
			isVoid: () => false,
			isInline: () => false
		}

		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = mount(<ListDropper type="ordered" bullets={bullets} editor={editor} />)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(editor.changeToType).toHaveBeenCalled()
	})

	test('List Dropper node opens and closes', () => {
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = mount(<ListDropper type="unordered" bullets={bullets} />)

		component
			.find('button')
			.at(1)
			.simulate('click')

		expect(component.html()).toMatchSnapshot()

		component
			.find('button')
			.at(1)
			.simulate('click')

		expect(component.html()).toMatchSnapshot()
	})

	test('List Dropper changes bullet', () => {
		const editor = {
			changeToType: jest.fn()
		}
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = mount(<ListDropper type="unordered" bullets={bullets} editor={editor} />)

		component
			.find('button')
			.at(2)
			.simulate('click')

		expect(editor.changeToType).toHaveBeenCalled()
	})

	test('List Dropper component opens and closes menu with keys', () => {
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = mount(<ListDropper type="unordered" bullets={bullets} />)

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowRight',
				stopPropagation: jest.fn()
			})

		expect(component.html()).toMatchSnapshot()

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowLeft',
				stopPropagation: jest.fn()
			})

		expect(component.html()).toMatchSnapshot()
	})

	test('List Dropper component moves up and down with keys', () => {
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = shallow(<ListDropper type="unordered" bullets={bullets} />)

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowUp',
				stopPropagation: jest.fn()
			})

		expect(component.state()).toMatchSnapshot()

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowDown',
				stopPropagation: jest.fn()
			})

		expect(component.state()).toMatchSnapshot()
	})

	test('List Dropper component closes menu when unfocused', () => {
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = shallow(<ListDropper type="unordered" bullets={bullets} />)

		const html = component
			.find('div')
			.at(0)
			.simulate('blur')
			.html()

		jest.runAllTimers()

		expect(html).toMatchSnapshot()
	})

	test('List Dropper component cancels menu closure when focused', () => {
		navigator.__defineGetter__('platform', () => 'Mock Mac')
		const bullets = [{ bulletStyle: 'disc', display: '●' }, { bulletStyle: 'circle', display: '○' }]
		const component = shallow(<ListDropper type="unordered" bullets={bullets} />)

		const html = component
			.find('div')
			.at(0)
			.simulate('focus')
			.html()

		jest.runAllTimers()

		expect(html).toMatchSnapshot()
	})
})
