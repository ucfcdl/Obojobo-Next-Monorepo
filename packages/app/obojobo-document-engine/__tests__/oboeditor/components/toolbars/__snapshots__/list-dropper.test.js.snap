// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Dropper List Dropper component cancels menu closure when focused 1`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-not-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-not-open\\"><button>●</button><button>○</button></div></div>"`;

exports[`List Dropper List Dropper component closes menu when unfocused 1`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-not-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-not-open\\"><button>●</button><button>○</button></div></div>"`;

exports[`List Dropper List Dropper component moves up and down with keys 1`] = `
Object {
  "currentFocus": NaN,
  "isOpen": false,
}
`;

exports[`List Dropper List Dropper component moves up and down with keys 2`] = `
Object {
  "currentFocus": NaN,
  "isOpen": false,
}
`;

exports[`List Dropper List Dropper component opens and closes menu with keys 1`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-open\\"><button>●</button><button>○</button></div></div>"`;

exports[`List Dropper List Dropper component opens and closes menu with keys 2`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-not-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-not-open\\"><button>●</button><button>○</button></div></div>"`;

exports[`List Dropper List Dropper node 1`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-not-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-not-open\\"><button>●</button><button>○</button></div></div>"`;

exports[`List Dropper List Dropper node opens and closes 1`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-open\\"><button>●</button><button>○</button></div></div>"`;

exports[`List Dropper List Dropper node opens and closes 2`] = `"<div class=\\"list-dropper\\" contenteditable=\\"false\\"><button class=\\"icon\\" title=\\"unordered list\\" aria-label=\\"unordered list\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"><path d=\\"M9.8,8.5c0,0.8-0.6,1.4-1.4,1.4S7,9.3,7,8.5s0.6-1.4,1.4-1.4S9.8,7.7,9.8,8.5z M9.8,16c0,0.8-0.6,1.4-1.4,1.4S7,16.8,7,16 c0-0.8,0.6-1.4,1.4-1.4S9.8,15.2,9.8,16z M9.8,23.5c0,0.8-0.6,1.4-1.4,1.4S7,24.3,7,23.5s0.6-1.4,1.4-1.4S9.8,22.7,9.8,23.5z\\"></path><rect x=\\"12.5\\" y=\\"8\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"23\\" width=\\"12.5\\" height=\\"1\\"></rect><rect x=\\"12.5\\" y=\\"15\\" width=\\"12.5\\" height=\\"1\\"></rect></svg></button><button class=\\"dropdown  is-not-open\\" aria-label=\\"Open unordered list type menu\\">⌃</button><div class=\\"list-dropper-menu  is-not-open\\"><button>●</button><button>○</button></div></div>"`;
